### PR TITLE
[flant-integration] Add metrics with raw node counters

### PR DESCRIPTION
**Pull request by shvgn**
## Description

Adding new metrics for nodes to be used in pricing telemetry.

## Why do we need it, and what problem does it solve?

Now we have metric `generate_node_count_metric` that is calculated right in the cluster. In case of
taint issues on control plane nodes, it contains wrong data. To address that, we have to fix the data
in the telemetry storage which requires investigation in every particular case and is inconvenient 
in general.

New metrics split the knowledge about nodes into independent pieces. They always have correct
raw data that will not be a subject for correction. The calculation of billed nodes will become this:

```promql
flant_pricing_nodes{type="ephemeral"} - flant_pricing_controlplane_tainted_nodes{type="ephemeral"}
flant_pricing_nodes{type="vm"}        - flant_pricing_controlplane_tainted_nodes{type="vm"}
flant_pricing_nodes{type="hard"}      - flant_pricing_controlplane_tainted_nodes{type="hard"}
flant_pricing_nodes{type="special"}   - flant_pricing_controlplane_tainted_nodes{type="special"}
```

## What is the expected result?

New raw metrics are being sent to the telemetry storage

## Checklist
- [x] Changes were tested in the Kubernetes cluster manually.

## Changelog entries

```changes
section: flant-integration
type: feature
summary: Added more node metrics to address issues with the billing for control-plane nodes without expected taints
```

### CODE REVIEW

File: ee/modules/600-flant-integration/images/flant-pricing/hooks/node_metrics.yaml [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+                  ]
+                  | length == 0
+              )
+          )
+      )
+      | {
+          "nodeGroup":        .metadata.labels."node.deckhouse.io/group",
+          "pricingNodeType": (.metadata.annotations."pricing.flant.com/nodeType"       // "unknown"),
+          "virtualization":  (.metadata.annotations."node.deckhouse.io/virtualization" // "unknown")
+      }
```
Comment (diafour) [inline ee/modules/600-flant-integration/images/flant-pricing/hooks/node_metrics.yaml:41]: 4 equal bindings look weird. It is possible to convert expressions in selects to boolean flags in output object and have only one binding for Nodes.

```
kubernetes:
  - name: nodes
    group: main
    kind: Node
    ...
    jqFilter: |
      {
          "nodeGroup":        .metadata.labels."node.deckhouse.io/group",
          "pricingNodeType": (.metadata.annotations."pricing.flant.com/nodeType"       // "unknown"),
          "virtualization":  (.metadata.annotations."node.deckhouse.io/virtualization" // "unknown"),
          "notControlPlaneNode": .metadata.labels."node.deckhouse.io/group" != "master" or ... or ... ,
          "masterNode": .metadata.labels."node.deckhouse.io/group" == "master",
          "taintedControlPlaneNode": .metadata.labels."node.deckhouse.io/group" == "master"    and ... and ... ,
          ...
      }
```

Comment (shvgn) [reply ee/modules/600-flant-integration/images/flant-pricing/hooks/node_metrics.yaml:41]: Sounds reasonable. I'll optimize it.

Comment (shvgn) [reply ee/modules/600-flant-integration/images/flant-pricing/hooks/node_metrics.yaml]: Done

File: ee/modules/600-flant-integration/images/flant-pricing/hooks/node_metrics.yaml [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+                          .spec.taints[]
+                          | select(
+                              .key == "node-role.kubernetes.io/control-plane" or
+                              .key == "node-role.kubernetes.io/master"
+                          )
+                      ]
+                      | length == 0
+                  )
+              )
+          ),
```
Comment (diafour) [inline ee/modules/600-flant-integration/images/flant-pricing/hooks/node_metrics.yaml:40]:
```suggestion
          "is_legacy_counted": (
              .metadata.labels."node.deckhouse.io/group" != "master"
              or
              .spec.taints == null
              or
              (
                  .spec.taints
                  | map(select(
                              .key == "node-role.kubernetes.io/control-plane" or
                              .key == "node-role.kubernetes.io/master"
                  )) | length == 0
              )
          ),
```

`map` is a shortcut for `[ .[] | x ]`

File: ee/modules/600-flant-integration/images/flant-pricing/hooks/node_metrics.yaml [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+                  [
+                      .spec.taints[]
+                      | select(
+                          .key == "node-role.kubernetes.io/control-plane" or
+                          .key == "node-role.kubernetes.io/master"
+                      )
+                  ]
+                  | length > 0
+              )
+          )
```
Comment (diafour) [inline ee/modules/600-flant-integration/images/flant-pricing/hooks/node_metrics.yaml:58]:
```suggestion
          "is_controlplane_tainted": (
              .metadata.labels."node.deckhouse.io/group" == "master"
              and
              .spec.taints != null
              and
              (
                  .spec.taints
                  | map(select(
                      .key == "node-role.kubernetes.io/control-plane" or
                      .key == "node-role.kubernetes.io/master"
                  )) | length > 0
              )
          )
```

File: ee/modules/600-flant-integration/images/flant-pricing/hooks/node_metrics.yaml (hunk 2 above)
Comment (shvgn) [reply ee/modules/600-flant-integration/images/flant-pricing/hooks/node_metrics.yaml:40]: I'd stick with current code. I don't see the value of this change, more of tiny doubts.

- I believe this syntax is less common
- In python list comprehension is more idiomatic than `map`, while writing in python it's more convenient to keep this construction
- I'd expect `map` to preserve the length of the array, so it's a little less intuitive. 



```shell
❯ jq  'map(select(.x!=null))' <<< '[{"x":1},{},{},{}]'
[
  {
    "x": 1
  }
]

❯ jq  '[ .[] | select(.x!=null) ]' <<< '[{"x":1},{},{},{}]'
[
  {
    "x": 1
  }
]
```

Comment (diafour) [reply ee/modules/600-flant-integration/images/flant-pricing/hooks/node_metrics.yaml:40]: jq magic =)

Comment (diafour) [inline ee/modules/600-flant-integration/images/flant-pricing/hooks/node_metrics.yaml:40]:
```suggestion
          "is_legacy_counted": (
              .metadata.labels."node.deckhouse.io/group" != "master"
              or
              .spec.taints == null
              or
              (
                  [
                      .spec.taints[]
                      | select(
                          .key == "node-role.kubernetes.io/control-plane" or
                          .key == "node-role.kubernetes.io/master"
                      )
                  ]
                  | length == 0
              )
          ),
```

At least, these `or` can be unindented

**Review by diafour — DISMISSED** [review]

**Review by diafour — DISMISSED** [review]

**Pull request merged**